### PR TITLE
arch/sim: add sim audio support

### PR DESCRIPTION
**Pull request by GUIDINGLI**
## Summary

arch/sim: add sim audio support

## Impact

sim audio will call alsa to playback/capture data

## Testing

nxplayer
nxrecoder

### CODE REVIEW

**Comment by acassis** [comment]
Very nice!

**Comment by masayuki2009** [comment]
@GUIDINGLI 
Could you please add a defconfig file so that we can try audio on sim?

**Comment by btashton** [comment]
I probably won't be able to really look at this until the weekend, but if someone wants to review and merge ahead of that I can look at it in master then.

File: arch/sim/src/sim/up_initialize.c [hunk 1]
```diff
@@ -281,4 +282,9 @@ void up_initialize(void)
 #if defined(CONFIG_FS_SMARTFS) && (defined(CONFIG_SIM_SPIFLASH) || defined(CONFIG_SIM_QSPIFLASH))
   up_init_smartfs();
 #endif
+
+#ifdef CONFIG_AUDIO
+  audio_register("pcm0p", sim_audio_initialize(true));
```
Comment (btashton) [inline arch/sim/src/sim/up_initialize.c:287]: Lets set these via `CONFIG_SIM_ALSA_DEVICE`

**Comment by acassis** [comment]
> @GUIDINGLI
> Could you please add a defconfig file so that we can try audio on sim?

In fact, this is a good idea to have a sim:audio example in place.

**Comment by btashton** [comment]
I'll update the test repo to be able to support this when I do my testing. We will need to have the alsa libraries available.

**Review by xiaoxiang781216 — COMMENTED** [review]
@GUIDINGLI please take @acassis and @btashton comment and update the patch.

**Comment by btashton** [comment]
@xiaoxiang781216 should I go ahead and get the ALSA libraries/headers in the CI images?  I can also reserve a name for the sim with an exception for Darwin.  That way this can run against CI, I don't see any reason this wont eventually make it in.

**Comment by xiaoxiang781216** [comment]
Sure we need install
sudo apt-get -y install libpulse-dev libpulse-dev:i386
sudo apt-get -y install libasound2-dev libasound2-dev:i386
sudo apt-get -y install libasound2-plugins libasound2-plugins:i386

**Comment by btashton** [comment]
> Sure we need install
> sudo apt-get -y install libpulse-dev libpulse-dev:i386
> sudo apt-get -y install libasound2-dev libasound2-dev:i386
> sudo apt-get -y install libasound2-plugins libasound2-plugins:i386

I'll do this tomorrow when do a quick verification on my hardware.

**Comment by btashton** [comment]
@xiaoxiang781216 I went ahead and just added it.  Note I only included `libasound2-dev libasound2-dev:i386` as it is all that should be required for building.  Adding pulseaudio pulls in a lot of deps so I hope that we can avoid that so we only add less than 10MB instead of 200MB+

@GUIDINGLI The config name I reserved is `alsa` so if you use that the CI should work for you.

File: arch/sim/src/Makefile [hunk 2]
```diff
@@ -202,6 +202,11 @@ ifeq ($(CONFIG_RPTUN),y)
   CSRCS += up_rptun.c
 endif
 
+ifeq ($(CONFIG_AUDIO),y)
```
[the remark this answers is not in the document]
Comment (GUIDINGLI) [reply arch/sim/src/Makefile]: done

File: arch/sim/src/sim/up_audio.c [hunk 3]
```diff
@@ -0,0 +1,590 @@
+/****************************************************************************
+ * arch/sim/src/sim/up_audio.c
```
[the remark this answers is not in the document]
Comment (GUIDINGLI) [reply arch/sim/src/sim/up_audio.c]: done

File: arch/sim/src/sim/up_initialize.c (hunk 1 above)
Comment (GUIDINGLI) [reply arch/sim/src/sim/up_initialize.c:287]: done

**Comment by acassis** [comment]
@btashton could you please verify if everything were fixed?

File: arch/sim/src/sim/up_initialize.c (hunk 1 above)
Comment (btashton) [reply arch/sim/src/sim/up_initialize.c:287]: @GUIDINGLI I don't see this in the recent commits

File: arch/sim/src/sim/up_internal.h [hunk 4]
```diff
@@ -397,6 +397,13 @@ int bthcisock_register(int dev_id);
 int bthcisock_loop(void);
 #endif
 
+/* up_audio.c ***************************************************************/
+
+#ifdef CONFIG_AUDIO
```
Comment (xiaoxiang781216) [inline arch/sim/src/sim/up_internal.h:402]: should change to CONFIG_SIM_SOUND

**Comment by masayuki2009** [comment]
@GUIDINGLI @xiaoxiang781216 
I tried this PR with ubuntu18.04 x86_64, but nuttx crashed when playback an WAV file via hostfs.
Could you tell me what environment are you using?

```
Thread 1 "nuttx" received signal SIGSEGV, Segmentation fault.                                                                                                                   
0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                        
87        inode = filep->f_inode;                                                                                                                                                                                                                                                                                                                     
p filep                                                                                                                                                                         
$1 = (struct file *) 0x10000000d8                                                                                                                                               
(gdb) p *filep                                                                                                                                                                  
Cannot access memory at address 0x10000000d8                                                                                                                                    
(gdb) up                                                                                                                                                                        
#1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
173           ret = file_vioctl(filep, req, ap);                                                                                                                                
(gdb) where                                                                                                                                                                     
#0  0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                    
#1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
#2  0x000055555557197a in ioctl (fd=4, req=4100) at vfs/fs_ioctl.c:259                                                                                                          
#3  0x000055555558f487 in nxplayer_setvolume (pplayer=0x7ffff2bfa0f0, volume=400) at nxplayer.c:1226                                                                            
#4  0x000055555558f164 in nxplayer_playthread (pvarg=0x7ffff2bfa0f0) at nxplayer.c:957                                                                                          
#5  0x0000555555583970 in pthread_start () at pthread/pthread_create.c:192                                                                                                      
#6  0x0000000000000000 in ?? ()     
```

**Comment by GUIDINGLI** [comment]
> @GUIDINGLI @xiaoxiang781216
> I tried this PR with ubuntu18.04 x86_64, but nuttx crashed when playback an WAV file via hostfs.
> Could you tell me what environment are you using?
> 
> ```
> Thread 1 "nuttx" received signal SIGSEGV, Segmentation fault.                                                                                                                   
> 0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                        
> 87        inode = filep->f_inode;                                                                                                                                                                                                                                                                                                                     
> p filep                                                                                                                                                                         
> $1 = (struct file *) 0x10000000d8                                                                                                                                               
> (gdb) p *filep                                                                                                                                                                  
> Cannot access memory at address 0x10000000d8                                                                                                                                    
> (gdb) up                                                                                                                                                                        
> #1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
> 173           ret = file_vioctl(filep, req, ap);                                                                                                                                
> (gdb) where                                                                                                                                                                     
> #0  0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                    
> #1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
> #2  0x000055555557197a in ioctl (fd=4, req=4100) at vfs/fs_ioctl.c:259                                                                                                          
> #3  0x000055555558f487 in nxplayer_setvolume (pplayer=0x7ffff2bfa0f0, volume=400) at nxplayer.c:1226                                                                            
> #4  0x000055555558f164 in nxplayer_playthread (pvarg=0x7ffff2bfa0f0) at nxplayer.c:957                                                                                          
> #5  0x0000555555583970 in pthread_start () at pthread/pthread_create.c:192                                                                                                      
> #6  0x0000000000000000 in ?? ()     
> ```

Could you show me your nxplayer cmd line ?

mine :
nsh> mount -t hostfs -o fs=/home/ligd/platform/FFmpeg/st /data
nsh> nxplayer
nxplayer> device /dev/audio/pcm0p
nxplayer> playraw /data/test2.wav 1 16 48000

**Comment by masayuki2009** [comment]
> > I tried this PR with ubuntu18.04 x86_64, but nuttx crashed when playback an WAV file via hostfs.
> > Could you tell me what environment are you using?
> > ```
> > Thread 1 "nuttx" received signal SIGSEGV, Segmentation fault.                                                                                                                   
> > 0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                        
> > 87        inode = filep->f_inode;                                                                                                                                                                                                                                                                                                                     
> > p filep                                                                                                                                                                         
> > $1 = (struct file *) 0x10000000d8                                                                                                                                               
> > (gdb) p *filep                                                                                                                                                                  
> > Cannot access memory at address 0x10000000d8                                                                                                                                    
> > (gdb) up                                                                                                                                                                        
> > #1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
> > 173           ret = file_vioctl(filep, req, ap);                                                                                                                                
> > (gdb) where                                                                                                                                                                     
> > #0  0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87                                                                    
> > #1  0x0000555555571682 in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:173                                                                                   
> > #2  0x000055555557197a in ioctl (fd=4, req=4100) at vfs/fs_ioctl.c:259                                                                                                          
> > #3  0x000055555558f487 in nxplayer_setvolume (pplayer=0x7ffff2bfa0f0, volume=400) at nxplayer.c:1226                                                                            
> > #4  0x000055555558f164 in nxplayer_playthread (pvarg=0x7ffff2bfa0f0) at nxplayer.c:957                                                                                          
> > #5  0x0000555555583970 in pthread_start () at pthread/pthread_create.c:192                                                                                                      
> > #6  0x0000000000000000 in ?? ()     
> > ```
> 
> @masayuki2009
> 
> Could you show me your nxplayer cmd line ?
> 
> mine :
> nsh> mount -t hostfs -o fs=/home/ligd/platform/FFmpeg/st /data
> nsh> nxplayer
> nxplayer> device /dev/audio/pcm0p
> nxplayer> playraw /data/test2.wav 1 16 48000

@GUIDINGLI 
The followings are actual sequences.

```
NuttShell (NSH) NuttX-8.2.0                                                                                                                               
nsh> mount -t hostfs -o fs=/home/ishikawa/nuttx_root /host                                                                                                
mount -t hostfs -o fs=/home/ishikawa/nuttx_root /host                                                                                                     
nsh> ls -l /host                                                                                                                                          
ls -l /host                                                                                                                                               
/host:                                                                                                                                                    
 drwxrwxr-x    4096 ..                                                                                                                                    
 -rw-rw-r--34108636 sample1.wav                                                                                                                           
 drwxrwxr-x    4096 .                                                                                                                                     
nsh> ls -l /dev/audio                                                                                                                                     
ls -l /dev/audio                                                                                                                                          
/dev/audio:                                                                                                                                               
 crw-rw-rw-       0 pcm0c                                                                                                                                 
 crw-rw-rw-       0 pcm0p                                                                                                                                 
nsh> nxplayer                                                                                                                                             
nxplayer                                                                                                                                                  
NxPlayer version 1.05                                                                                                                                     
h for commands, q to exit                                                                                                                                 
                                                                                                                                                          
nxplayer> device /dev/audio/pcm0p                                                                                                                         
device /dev/audio/pcm0p                                                                                                                                   
nxplayer> playraw /host/sample1.wav 2 16 44100  
```

**Comment by GUIDINGLI** [comment]
@masayuki2009

Have you pick another patch for increase stack size ?

https://github.com/apache/incubator-nuttx/pull/2124/commits/d86d7bc177d9180022549d4d4476b88127510f79

From your log:
#0  0x00005555555714a2 in file_vioctl (filep=0x10000000d8, req=4100, ap=0x7ffff2c8b520) at vfs/fs_ioctl.c:87

filep address is invaild

in my gdb:

filep is the same area with ap, both 0x7ffff2c...

#0  file_vioctl (filep=0x7ffff2c7a8c8, req=4100, ap=0x7ffff2c9af40) at vfs/fs_ioctl.c:87
#1  0x000055555556fb4c in nx_vioctl (fd=4, req=4100, ap=0x7ffff2c9af40) at vfs/fs_ioctl.c:173
#2  0x000055555556fe44 in ioctl (fd=4, req=4100) at vfs/fs_ioctl.c:259
#3  0x000055555558cc96 in nxplayer_setvolume (pplayer=0x7ffff2bfa0f0, volume=400) at nxplayer.c:1226
#4  0x000055555558c973 in nxplayer_playthread (pvarg=0x7ffff2bfa0f0) at nxplayer.c:957
#5  0x0000555555581e3a in pthread_start () at pthread/pthread_create.c:192
#6  0x0000000000000000 in ?? ()

**Comment by masayuki2009** [comment]
>Have you pick another patch for increase stack size ?

@GUIDINGLI 
Yes, because I checked out your repository and branch.

**Comment by GUIDINGLI** [comment]
@masayuki2009 

You should also update apps,:

https://github.com/apache/incubator-nuttx-apps/pull/461

**Comment by masayuki2009** [comment]
>You should also update apps,:
>
>apache/incubator-nuttx-apps#461

@GUIDINGLI 
Thanks, it works now!

**Comment by masayuki2009** [comment]
> @xiaoxiang781216 I went ahead and just added it. Note I only included `libasound2-dev libasound2-dev:i386` as it is all that should be required for building. Adding pulseaudio pulls in a lot of deps so I hope that we can avoid that so we only add less than 10MB instead of 200MB+
> 
> @GUIDINGLI The config name I reserved is `alsa` so if you use that the CI should work for you.

@btashton @xiaoxiang781216 @GUIDINGLI 
I only installed libasound2-dev onto my ubuntu18.04 x86_64 machine.

**Comment by acassis** [comment]
Hi @masayuki2009 could you please give an Approved from your side? If nobody is against it we can merge it now!

**Comment by btashton** [comment]
No. My comment has not been addressed.

**Comment by xiaoxiang781216** [comment]
@btashton could you review again? I think your comment is already addressed in the latest revision.

File: arch/sim/src/sim/up_initialize.c (hunk 1 above)
Comment (btashton) [reply arch/sim/src/sim/up_initialize.c:287]: @xiaoxiang781216 still hard coded here or am I not understanding

File: arch/sim/src/sim/up_alsa.c [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+
+  if (priv->pcm)
+    {
+      return 0;
+    }
+
+  direction = priv->playback ? SND_PCM_STREAM_PLAYBACK
+                             : SND_PCM_STREAM_CAPTURE;
+
+  ret = snd_pcm_open(&pcm, "default", direction, 0);
```
Comment (btashton) [inline arch/sim/src/sim/up_alsa.c:221]: Ah I guess this is actually where my comment should be. Don't we want to specify this?

Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_alsa.c:221]: Do you want to change the device name? Ok, it depend on how many pcm instances we want to create. only one playback/capture is created in the current design, "default" should be the best choice in this case. On the other hand, we sim_audio_initialize need accept name argument and open the different sound device.

Comment (btashton) [reply arch/sim/src/sim/up_alsa.c:221]: I guess I don't care if we limit to one device, but I want to be able to at least specify it via a config. "default" is certainly the correct default choice.

Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_alsa.c:221]: In ALSA design, user can modify alsa.conf to map "default" to the different sound hardware. That's why we don't make the device name configurable initially. If you really need this feature, we can add it.

Comment (btashton) [reply arch/sim/src/sim/up_alsa.c:221]: Just seems trivial to add the two kconfig values here and then not require people to modify their system for testing.  This makes it a lot easier to use an alsa loopback device.

Comment (btashton) [reply arch/sim/src/sim/up_alsa.c:221]: If you want to still merge fine, it just does not align with other sim hardware where the user can specify how it connects to to the host hardware.

Comment (xiaoxiang781216) [reply arch/sim/src/sim/up_alsa.c:221]: How about we merge this patch first? and  @GUIDINGLI provide a new patch to support configuration in tomorrow.

**Pull request merged**